src/tls/ffi.c

```diff
@@ -0,0 +1,402 @@
+    } else {
+        mitls_state * s;
+        
+        // Allocate space on the heap, to store an OCaml value
+        s = (mitls_state*)malloc(sizeof(mitls_state));
+        if (s) {
+            // Tell the OCaml GC about the heap address, so it is treated
+            // as a GC root, keeping the config object live.
+            s->fstar_state = config; 
+            caml_register_generational_global_root(&s->fstar_state);
+            if (strcmp(tls_version, "1.3") == 0) {
+                s->connect_callback = g_mitls_FFI_Connect13;
+            } else {
+                s->connect_callback = g_mitls_FFI_Connect12;
+            }
+            *state = s;
+            ret = 1;
+        }
+    }
+    
+    restore_stdio(&r, outmsg, errmsg);
+    return ret;
+}
+
+// Called by the host app to free a mitls_state allocated by FFI_mitls_configure()
+void FFI_mitls_close(mitls_state *state)
+{
+    if (state) {
+        caml_remove_generational_global_root(&state->fstar_state);
+        state->fstar_state = 0;
+        free(state);
+    }
+}
+
+void FFI_mitls_free_msg(char *msg)
+{
+    free(msg);
+}
+
+void FFI_mitls_free_packet(void *packet)
+{
+    free(packet);
+}
+
+void * copypacket(value packet, /* out */ size_t *packet_size)
+{
+    void *p;
+    mlsize_t size;
+        
+    size = caml_string_length(packet);
+    p = malloc(size);
+    if (p) {
+        memcpy(p, String_val(packet), size);
+        *packet_size = size;
+    }
+    return p;
+}
+
+// Called from FStar code to send via TCP
+CAMLprim value ocaml_send_tcp(value cookie, value bytes)
+{
+    mlsize_t buffer_size;
+    char *buffer;
+    int retval;
+    size_t c = Long_val(cookie);
+    struct _FFI_mitls_callbacks *callbacks = (struct _FFI_mitls_callbacks *)c;
+    CAMLparam2(cookie, bytes);
+
+    buffer = Bp_val(bytes);
+    buffer_size = caml_string_length(bytes);
+    
+    retval = (*callbacks->send)(callbacks, buffer, buffer_size);
+    
+    return Val_int(retval);
+}
+
+// Called from FStar code to receive via TCP
+CAMLprim value ocaml_recv_tcp(value cookie, value bytes)
+{
+    mlsize_t buffer_size;
+    char *buffer;
+    ssize_t retval;
+    size_t c = Long_val(cookie);
+    struct _FFI_mitls_callbacks *callbacks = (struct _FFI_mitls_callbacks *)c;
+    CAMLparam2(cookie, bytes);
+    
+    buffer = Bp_val(bytes);
+    buffer_size = caml_string_length(bytes);
+    
+    retval = (*callbacks->recv)(callbacks, buffer, buffer_size);
+    
+    return Val_int(retval);
+}
+
+// Called by the host app to create a TLS connection.
+int FFI_mitls_connect(struct _FFI_mitls_callbacks *callbacks, /* in */ mitls_state *state, /* out */ char **outmsg, /* out */ char **errmsg)
+{
+    CAMLlocal1(result);
+    int ret;
+    mitls_redirect r;
+    
+    *outmsg = NULL;
+    *errmsg = NULL;
+    redirect_stdio(&r);
+    
+    result = caml_callback2_exn(*state->connect_callback, state->fstar_state, Val_long((size_t)callbacks));
```

Comment:
Correct - there will be a different _FFI_mitls_callbacks pointer for each TLS connection that curl initiates.  The two function pointers inside it are always the same, but the address of the struct, and the contents of the rest of the memory around those pointers, will change.  The C code in those function pointers knows how to take the callbacks address and reconstruct the rest of the state variables.  That includes the socket handle, curl logging buffers, etc.

So no, passing an enum representing the callback isn't sufficient to complete the call.